src/res/mod.rs

```diff
@@ -244,124 +338,297 @@ impl Resources {
     /// let mut res = Resources::new();
     /// res.add(MyRes(5));
     /// ```
-    pub fn add<R>(&mut self, r: R)
+    pub fn add<R>(&mut self, r: R) -> Option<R>
     where
         R: Resource,
     {
         use std::collections::hash_map::Entry;
 
+        self.maintain();
+
         let entry = self.resources.entry(ResourceId::new::<R>());
 
         if let Entry::Vacant(e) = entry {
             e.insert(TrustCell::new(Box::new(r)));
+
+            None
         } else {
-            panic!("Tried to add a resource though \
-                    an instance of this type already exists in `Resources`");
+            Some(r)
+        }
+    }
+
+    /// Adds a resource or panics if it already exists.
+    /// This was the old behavior of `add`.
+    ///
+    /// Calls `maintain` at the beginning.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the resource has already been added.
+    ///
+    /// # Other variants
+    ///
+    /// See `add` for the non-panicking version.
+    /// See `add_or_overwrite` for a version that just assures a resource has a certain value.
+    pub fn add_assert<R>(&mut self, r: R)
```

Comment:
This is a bad name, suggestions welcome.

Reply:
`add_no_overwrite`?